### ISSUE TITLE
Pre-render certain routes during build

### ISSUE THREAD

**Issue by sebastianmacias** [lukeed/inferno-starter#3]
First of all, I love how clean and lean this starter is. Great work!

Since the idea behind this starter is to use it as starting point for static/off-line first apps it would be amazing if certain routes could be pre-rendered and added to the dist folder during build.

For instance /blog, /blog/article1, etc.

That way the end user could upload the dist folder to S3 or any static host and routes like /blog/article1 could be accessed directly without having to visit the homepage first. This will allow sharing deep links and will help with SEO as well.

**Comment by lukeed** [comment]
Thanks @sebastianmacias! 

Well, with the current setup, the end user only has to run `npm start` and all pages will serve the original `index.html`. I really don't want to include any SSR or server-based solutions beyond that. I, for one, don't need that in 95% of my projects.

**Comment by lukeed** [comment]
What do you think about that? Is `npm start` enough?

**Comment by sebastianmacias** [comment]
Yes, it will be enough for lots of use cases (but all of them would require a server with a nodejs procces running just to take care of the front-end side of an app).

I actually prefer to have my front-end 100% statically served and not to worry about scaling that part of my apps. I let S3 and/or a CDN take care of that.

All my backend logic runs (lately) on AWS lambda so basically it's all "serverless'.

That's my use case but, still if you allow route prerendering using the provided methods:

renderToString(infernoComponent) 
renderToStaticMarkup(infernoComponent) 

this starter project could easily be used as an static generator and that would cover a lot more use cases.

There is a growing tendency towards statically generated sites, that's for sure.

What do you think?

**Comment by lukeed** [comment]
Interesting. I just fear it'll overly complicate the boilerplate setup. I really wanted this to be beginner friendly; overly complex setups and commands scare new & old devs who are approaching any library. 

I should mention that I am building a more advanced / SSR capable kit, and maybe this will be more applicable there?

Either way, you're looking to run something like `npm run render` and produce additional content:

```
dist
  |- index.html
  |- manifest.jsom
  |- etc...
  |- img/
  |- blog/
    |- index.html 
    |- article1.html
```

And then a `.htaccess` for extension rewrites?

**Comment by sebastianmacias** [comment]
Yes, but instead of producing:

dist
  |- index.html
  |- manifest.jsom
  |- etc...
  |- img/
  |- blog/
    |- index.html 
    |- article1.html

It would produce:

dist
  |- index.html
  |- manifest.jsom
  |- etc...
  |- img/
  |- blog/
    |- index.html 
    |- article1/
         |-  index.html

So no .htaccess is required and it can be served from anywhere

**Comment by sebastianmacias** [comment]
Check out https://www.staticgen.com to get at feel of the demand there is for this kind of solution (off course without over complicating the scope of the project)

Not having to run 'npm start' to serve the site but instead 'npm build render' and then uploading the contents of 'dist' to a CDN and still be able to call microservices  (if you need to) from the static site would be amazing

**Comment by sebastianmacias** [comment]
Of course a pre rendered route like /blog/article1/ should be isomorphic. Once loaded from the static host the end user should be able to interact with the page just like if it was served by npm run

**Comment by lukeed** [comment]
Haven't forgot about this @sebastianmacias 😄 

You mentioned that "certain routes" would be rendered. How would one specify some routes/pages over others?

**Comment by sebastianmacias** [comment]
Hi @lukeed ,

I think package.json would be a good place to specify the isomorphic routes, for instance:

```
{
  "private": true,
  "name": "inferno-starter",
  "version": "0.1.1",
  "license": "MIT",
  "repository": "lukeed/inferno-starter",
  "author": {
    "name": "Luke Edwards",
    "email": "luke.edwards05@gmail.com",
    "url": "https://lukeed.com"
  },
  "scripts": {
    "build": "NODE_ENV=production webpack --config config/webpack",
    ...
  },
  "isomorphic-routes": {
    "/blog": null,
    "/blog/:title": ["article1", "article2", "article3"],
    "/products/:id": [1,2,3,4]
  },
  "dependencies": {
    "history": "^4.5.0",
    "inferno": "^1.1.1",
    ...
  },
  "devDependencies": {
    "autoprefixer": "^6.6.1",
      ...
  }
}
```
